### PR TITLE
Extract update_weights from RL Engine to SGLang to keep simplicity and fix torch reduce

### PR DESCRIPTION
**Pull request by hebiao064**
Co-authored-by: CuiBo [82354186+SuperCB@users.noreply.github.com](mailto:82354186+SuperCB@users.noreply.github.com)
Co-authored-by: GeLee [865038696@qq.com](mailto:865038696@qq.com)
Co-authored-by: 杨睿 <yangruipis@163.com>

## Context
Builds on the earlier work from [#2620](https://github.com/volcengine/verl/pull/2620) to further simplify and unify the update weights logic 

## Motivation

We've received numerous complaints that the update weights logic in VERL SGLang is difficult to understand and error-prone during development. For example, [a recent refactor from VerlEngine to AsyncEngine](https://github.com/hebiao064/verl/commit/fe23634116ab69cacc38e7570e3c0e344eb1c711) accidentally broke the torch reduce patch logic, leading to a 3× slowdown in weight updates.

Additionally, when changing the tensor update logic from per-tensor to batched updates, we had to duplicate the logic across both [Megatron SGLang Sharding Manager](https://github.com/volcengine/verl/pull/2418) and [FSDP SGLang Sharding Manager](https://github.com/volcengine/verl/pull/2559). This not only introduced unnecessary complexity, but also made maintenance and future improvements harder.

To address these issues, we propose simplifying and centralizing the update weights logic within core SGLang, exposing a unified API that is transparent to all RL engines. This abstraction reduces duplication, improves reliability, and accelerates future development.

### Benefits
- ✅ Unified and transparent update weights API for all RL engines (FSDP, Megatra, etc.)
- 🚀 Speedup QWen3 MoE weight updates from 110s to 30s
- 🔧 Torch patch MP serializer to avoid inefficient P2P communication that slows down weight updates

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Review by zhuzilin — APPROVED** [review]

File: python/sglang/srt/weight_sync/utils.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    Preprocess the tensor for update weights.
+    Example Use Case:
+        - FSDP: we gather tensor by calling full_tensor in _preprocess_tensor_for_update_weights
+        - Megatron: we do nothing here, assuming it is gathered when feed into this func
+
+    Args:
+        tensor: The tensor to be preprocessed.
+
+    Returns:
+        The full tensor.
```
Comment (fzyzcjy) [inline python/sglang/srt/weight_sync/utils.py]: not sure whether full tensor. iirc for megatron they said long ago it is exactly the tensor on this gpu needed by sgl, i.e. a slice of tensor.

File: python/sglang/srt/weight_sync/utils.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    Args:
+        engine: The inference engine created by the caller process.
+        params_batch: A list of (name, tensor) tuples. We batched the tensors to avoid the overhead of cpu call.
+        device_mesh_key: The key of the device mesh. Typically "tp" or "infer_tp"
+        device_mesh: The device mesh.
+        load_format: The format of the weights.
+    """
+    infer_tp_size = device_mesh[device_mesh_key].mesh.size()[0]
+    infer_tp_rank = device_mesh[device_mesh_key].get_local_rank()
+    # Monkey patching Torch to ensure torch reduce tensor
```
Comment (fzyzcjy) [inline python/sglang/srt/weight_sync/utils.py]: nit: I personally think we do not this many comments but that's also ok

**Review by fzyzcjy — COMMENTED** [review]
LGTM as long as the logic is directly moved from verl and no change to it

File: python/sglang/srt/weight_sync/utils.py (hunk 1 above)
Comment (hebiao064) [reply python/sglang/srt/weight_sync/utils.py]: yes for megatron its a full tensor already...

File: python/sglang/srt/weight_sync/utils.py (hunk 2 above)
Comment (hebiao064) [reply python/sglang/srt/weight_sync/utils.py]: removed

**Comment by hebiao064** [comment]
Util recent commit, FSDP + SGLang works fine with expected speed up:
` timing_s/reshard:28.928760528564453 `

@fzyzcjy I will run test against megatron as well

**Comment by zhaochenyang20** [comment]
Great! Let me keep track of this

File: python/sglang/srt/weight_sync/utils.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    # [
+    #   (name0, ipc_tensor0_tp0),
+    #   (name1, ipc_tensor1_tp0),
+    # ]
+    named_tensors_batch = [
+        (
+            name,
+            MultiprocessingSerializer.serialize(
+                _preprocess_tensor_for_update_weights(tensor)
```
Comment (hebiao064) [inline python/sglang/srt/weight_sync/utils.py:48]: I should do tensor.detach() here or in _preprocess_tensor_for_update_weights
